web/modules/custom/joinup_core/src/EventSubscriber/GroupInvitationSubscriber.php

```diff
@@ -0,0 +1,116 @@
+   *
+   * @param \Drupal\og\MembershipManagerInterface $membershipManager
+   *   The membership manager service.
+   */
+  public function __construct(MembershipManagerInterface $membershipManager) {
+    $this->membershipManager = $membershipManager;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function getSubscribedEvents() : array {
+    $events[InvitationEvents::ACCEPT_INVITATION_EVENT] = ['acceptInvitation'];
+    $events[InvitationEvents::REJECT_INVITATION_EVENT] = ['rejectInvitation'];
+
+    return $events;
+  }
+
+  /**
+   * Accepts invitations to join a group.
+   *
+   * @param \Drupal\joinup_invite\Event\InvitationEventInterface $event
+   *   The event that was fired.
+   */
+  public function acceptInvitation(InvitationEventInterface $event) : void {
+    $invitation = $event->getInvitation();
+
+    // Ignore invitations to other content entities.
+    if ($invitation->bundle() !== 'group') {
+      return;
+    }
+
+    $user = $invitation->getRecipient();
+    $group = $invitation->getEntity();
+
+    // We don't need to act if another action has been taken on the membership.
+    $membership = $this->membershipManager->getMembership($group, $user, [OgMembershipInterface::STATE_PENDING]);
+    if (empty($membership)) {
+      drupal_set_message($this->t('There is no action pending for this user.'));
+    }
+    else {
+      // Disable notifications related to memberships.
+      $membership->skip_notification = TRUE;
+      $membership->setState(OgMembershipInterface::STATE_ACTIVE)->save();
+      $facilitator_id = $group->getEntityTypeId() . '-' . $group->bundle() . '-facilitator';
+      $role_argument = $membership->hasRole($facilitator_id) ? 'facilitator' : 'member';
+      drupal_set_message($this->t('You are now a :role of the ":title" :bundle.', [
+        ':role' => $role_argument,
+        ':title' => $group->label(),
+        ':bundle' => $group->bundle(),
+      ]));
+    }
+  }
+
+  /**
+   * Rejects invitations to join a group.
+   *
+   * @param \Drupal\joinup_invite\Event\InvitationEventInterface $event
+   *   The event that was fired.
+   */
+  public function rejectInvitation(InvitationEventInterface $event) : void {
+    $invitation = $event->getInvitation();
+
+    // Ignore invitations to other content entities.
+    if ($invitation->bundle() !== 'group') {
+      return;
+    }
+
+    $user = $invitation->getRecipient();
+    $group = $invitation->getEntity();
+
+    // We don't need to act if another action has been taken on the membership.
+    $membership = $this->membershipManager->getMembership($group, $user, [OgMembershipInterface::STATE_PENDING]);
+    if (empty($membership)) {
+      drupal_set_message($this->t('There is no action pending for this user.'));
+    }
+    else {
+      // Disable notifications related to memberships.
+      $membership->skip_notification = TRUE;
+      // Deleting the membership will clear the invitation as well.
```

Comment:
That was also a tough part in my implementation (I don't know if I should document this somewhere).

The code in the `OgMembershipSubscriber::appliedOnDelete()` is because the user might be invited and have a pending membership, and then a facilitator deletes him using the bulk action. In that case, the membership is deleted without the notification being deleted.

As for the rest, I kinda like the approach with the hook_invitation_delete(). I will try it. The reason for this code is that we need to delete the invitation first as if the invitation is deleted first, we do not have a way to know if the pending membership was a result of an invitation or a membership request or any other reason.

I will test the hook though and let you know.

Reply:
Now that I am thinking of it, the `joinup_invite_og_membership_delete()` is also needed for the same case, where a membership is deleted from the UI and the invitation needs to be cleared too.

Reply:
I can tell that is something I don't like here conceptually: this the redundancy between a pending membership and an invitation. In this scenario, they have playing the same role. So, what is pending membership? I think the pending state of membership makes more sense when the user (candidate for membership) is willing to join a group and he applies for a membership. By applying he's granted with pending membership. Then the group admins will make the decision to approve or deny the candidacy. On the other hand, when inviting it's not the same thing. When a group admin invites someone he has already approved his membership. This is because when the recipient accepts the invitation he's active member, no admin action is required. So, IMO here we don't have to create a pending membership when we create an invitation. We're just creating the invitation entity and when the recipient accepts the invitation then we create the membership directly with the **active** status. If the user denies the invitation, we only have to drop the invitation entity and we're done. So we're avoiding that ugly redundancy and circular dependency. What you think?